// @syncfusion/ej2-vue-splitbuttons/src/split-button/items.directive.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { EJComponentDecorator, allVue, gh } from '@syncfusion/ej2-vue-base';

import Vue from 'vue';
export var isExecute = gh ? false : true;
var vueImport;
if (!isExecute || parseInt(allVue.version) < 3) {
    vueImport = Vue;
}
else {
    vueImport = Vue;
}
var SplitButtonItemsDirective = /** @class */ (function (_super) {
    __extends(SplitButtonItemsDirective, _super);
    function SplitButtonItemsDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SplitButtonItemsDirective.prototype.render = function () {
        return;
    };
    SplitButtonItemsDirective.prototype.getTag = function () {
        return 'e-splitbuttonitems';
    };
    SplitButtonItemsDirective = __decorate([
        EJComponentDecorator({}, isExecute)
    ], SplitButtonItemsDirective);
    return SplitButtonItemsDirective;
}(vueImport));
export { SplitButtonItemsDirective };
export var SplitButtonItemsPlugin = {
    name: 'e-splitbuttonitems',
    install: function (Vue) {
        Vue.component(SplitButtonItemsPlugin.name, SplitButtonItemsDirective);
    }
};
var SplitButtonItemDirective = /** @class */ (function (_super) {
    __extends(SplitButtonItemDirective, _super);
    function SplitButtonItemDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SplitButtonItemDirective.prototype.render = function () {
        return;
    };
    SplitButtonItemDirective.prototype.getTag = function () {
        return 'e-splitbuttonitem';
    };
    SplitButtonItemDirective = __decorate([
        EJComponentDecorator({}, isExecute)
    ], SplitButtonItemDirective);
    return SplitButtonItemDirective;
}(vueImport));
export { SplitButtonItemDirective };
export var SplitButtonItemPlugin = {
    name: 'e-splitbuttonitem',
    install: function (Vue) {
        Vue.component(SplitButtonItemPlugin.name, SplitButtonItemDirective);
    }
};
