// @syncfusion/ej2-vue-spreadsheet/src/spreadsheet/image.directive.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { EJComponentDecorator, allVue, gh } from '@syncfusion/ej2-vue-base';

import Vue from 'vue';
export var isExecute = gh ? false : true;
var vueImport;
if (!isExecute || parseInt(allVue.version) < 3) {
    vueImport = Vue;
}
else {
    vueImport = Vue;
}
var ImagesDirective = /** @class */ (function (_super) {
    __extends(ImagesDirective, _super);
    function ImagesDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ImagesDirective.prototype.render = function () {
        return;
    };
    ImagesDirective.prototype.getTag = function () {
        return 'e-images';
    };
    ImagesDirective = __decorate([
        EJComponentDecorator({}, isExecute)
    ], ImagesDirective);
    return ImagesDirective;
}(vueImport));
export { ImagesDirective };
export var ImagesPlugin = {
    name: 'e-images',
    install: function (Vue) {
        Vue.component(ImagesPlugin.name, ImagesDirective);
    }
};
var ImageDirective = /** @class */ (function (_super) {
    __extends(ImageDirective, _super);
    function ImageDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ImageDirective.prototype.render = function () {
        return;
    };
    ImageDirective.prototype.getTag = function () {
        return 'e-image';
    };
    ImageDirective = __decorate([
        EJComponentDecorator({}, isExecute)
    ], ImageDirective);
    return ImageDirective;
}(vueImport));
export { ImageDirective };
export var ImagePlugin = {
    name: 'e-image',
    install: function (Vue) {
        Vue.component(ImagePlugin.name, ImageDirective);
    }
};
