// @syncfusion/ej2-vue-spreadsheet/dist/es6/ej2-vue-spreadsheet.es2015.js
import { Spreadsheet } from '@syncfusion/ej2-spreadsheet';
import { ComponentBase, EJComponentDecorator, allVue, getProps, gh } from '@syncfusion/ej2-vue-base';
import { Vue } from 'vue-class-component';
import Vue$1 from 'vue';
import { isNullOrUndefined } from '@syncfusion/ej2-base';

var __decorate = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
const isExecute = gh ? false : true;
let vueImport;
if (!isExecute || parseInt(allVue.version) < 3) {
    vueImport = Vue;
}
else {
    vueImport = Vue$1;
}
let ImagesDirective = class ImagesDirective extends vueImport {
    render() {
        return;
    }
    getTag() {
        return 'e-images';
    }
};
ImagesDirective = __decorate([
    EJComponentDecorator({}, isExecute)
], ImagesDirective);
const ImagesPlugin = {
    name: 'e-images',
    install(Vue$$1) {
        Vue$$1.component(ImagesPlugin.name, ImagesDirective);
    }
};
let ImageDirective = class ImageDirective extends vueImport {
    render() {
        return;
    }
    getTag() {
        return 'e-image';
    }
};
ImageDirective = __decorate([
    EJComponentDecorator({}, isExecute)
], ImageDirective);
const ImagePlugin = {
    name: 'e-image',
    install(Vue$$1) {
        Vue$$1.component(ImagePlugin.name, ImageDirective);
    }
};

var __decorate$1 = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
const isExecute$1 = gh ? false : true;
let vueImport$1;
if (!isExecute$1 || parseInt(allVue.version) < 3) {
    vueImport$1 = Vue;
}
else {
    vueImport$1 = Vue$1;
}
let ChartsDirective = class ChartsDirective extends vueImport$1 {
    render() {
        return;
    }
    getTag() {
        return 'e-charts';
    }
};
ChartsDirective = __decorate$1([
    EJComponentDecorator({}, isExecute$1)
], ChartsDirective);
const ChartsPlugin = {
    name: 'e-charts',
    install(Vue$$1) {
        Vue$$1.component(ChartsPlugin.name, ChartsDirective);
    }
};
let ChartDirective = class ChartDirective extends vueImport$1 {
    render() {
        return;
    }
    getTag() {
        return 'e-chart';
    }
};
ChartDirective = __decorate$1([
    EJComponentDecorator({}, isExecute$1)
], ChartDirective);
const ChartPlugin = {
    name: 'e-chart',
    install(Vue$$1) {
        Vue$$1.component(ChartPlugin.name, ChartDirective);
    }
};

var __decorate$2 = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
const isExecute$2 = gh ? false : true;
let vueImport$2;
if (!isExecute$2 || parseInt(allVue.version) < 3) {
    vueImport$2 = Vue;
}
else {
    vueImport$2 = Vue$1;
}
let CellsDirective = class CellsDirective extends vueImport$2 {
    render() {
        return;
    }
    getTag() {
        return 'e-cells';
    }
};
CellsDirective = __decorate$2([
    EJComponentDecorator({}, isExecute$2)
], CellsDirective);
const CellsPlugin = {
    name: 'e-cells',
    install(Vue$$1) {
        Vue$$1.component(CellsPlugin.name, CellsDirective);
    }
};
/**
 * `e-cell` directive represent a cell of the VueJS Spreadsheet.
 * It must be contained in a `e-row` directive.
 * ```vue
 * <ejs-spreadsheet>
 *   <e-sheets>
 *    <e-sheet>
 *    <e-rows>
 *    <e-row>
 *    <e-cells>
 *    <e-cell value='A1'></e-cell>
 *    </e-cells>
 *    </e-row>
 *    </e-rows>
 *    </e-sheet>
 *   </e-sheets>
 * </ejs-spreadsheet>
 * ```
 */
let CellDirective = class CellDirective extends vueImport$2 {
    render() {
        return;
    }
    getTag() {
        return 'e-cell';
    }
};
CellDirective = __decorate$2([
    EJComponentDecorator({}, isExecute$2)
], CellDirective);
const CellPlugin = {
    name: 'e-cell',
    install(Vue$$1) {
        Vue$$1.component(CellPlugin.name, CellDirective);
    }
};

var __decorate$3 = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
const isExecute$3 = gh ? false : true;
let vueImport$3;
if (!isExecute$3 || parseInt(allVue.version) < 3) {
    vueImport$3 = Vue;
}
else {
    vueImport$3 = Vue$1;
}
let RowsDirective = class RowsDirective extends vueImport$3 {
    render() {
        return;
    }
    getTag() {
        return 'e-rows';
    }
};
RowsDirective = __decorate$3([
    EJComponentDecorator({}, isExecute$3)
], RowsDirective);
const RowsPlugin = {
    name: 'e-rows',
    install(Vue$$1) {
        Vue$$1.component(RowsPlugin.name, RowsDirective);
    }
};
/**
 * `e-row` directive represent a row of the VueJS Spreadsheet.
 * It must be contained in a `e-sheet` directive.
 * ```vue
 * <ejs-spreadsheet>
 *   <e-sheets>
 *    <e-sheet>
 *    <e-rows>
 *    <e-row></e-row>
 *    </e-rows>
 *    </e-sheet>
 *   </e-sheets>
 * </ejs-spreadsheet>
 * ```
 */
let RowDirective = class RowDirective extends vueImport$3 {
    render() {
        return;
    }
    getTag() {
        return 'e-row';
    }
};
RowDirective = __decorate$3([
    EJComponentDecorator({}, isExecute$3)
], RowDirective);
const RowPlugin = {
    name: 'e-row',
    install(Vue$$1) {
        Vue$$1.component(RowPlugin.name, RowDirective);
    }
};

var __decorate$4 = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
const isExecute$4 = gh ? false : true;
let vueImport$4;
if (!isExecute$4 || parseInt(allVue.version) < 3) {
    vueImport$4 = Vue;
}
else {
    vueImport$4 = Vue$1;
}
let ColumnsDirective = class ColumnsDirective extends vueImport$4 {
    render() {
        return;
    }
    getTag() {
        return 'e-columns';
    }
};
ColumnsDirective = __decorate$4([
    EJComponentDecorator({}, isExecute$4)
], ColumnsDirective);
const ColumnsPlugin = {
    name: 'e-columns',
    install(Vue$$1) {
        Vue$$1.component(ColumnsPlugin.name, ColumnsDirective);
    }
};
/**
 * `e-column` directive represent a column of the VueJS Spreadsheet.
 * It must be contained in a `e-sheet` directive.
 * ```vue
 * <ejs-spreadsheet>
 *   <e-sheets>
 *    <e-sheet>
 *    <e-columns>
 *    <e-column></e-column>
 *    </e-columns>
 *    </e-sheet>
 *   </e-sheets>
 * </ejs-spreadsheet>
 * ```
 */
let ColumnDirective = class ColumnDirective extends vueImport$4 {
    render() {
        return;
    }
    getTag() {
        return 'e-column';
    }
};
ColumnDirective = __decorate$4([
    EJComponentDecorator({}, isExecute$4)
], ColumnDirective);
const ColumnPlugin = {
    name: 'e-column',
    install(Vue$$1) {
        Vue$$1.component(ColumnPlugin.name, ColumnDirective);
    }
};

var __decorate$5 = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
const isExecute$5 = gh ? false : true;
let vueImport$5;
if (!isExecute$5 || parseInt(allVue.version) < 3) {
    vueImport$5 = Vue;
}
else {
    vueImport$5 = Vue$1;
}
let RangesDirective = class RangesDirective extends vueImport$5 {
    render() {
        return;
    }
    getTag() {
        return 'e-ranges';
    }
};
RangesDirective = __decorate$5([
    EJComponentDecorator({}, isExecute$5)
], RangesDirective);
const RangesPlugin = {
    name: 'e-ranges',
    install(Vue$$1) {
        Vue$$1.component(RangesPlugin.name, RangesDirective);
    }
};
/**
 * `e-range` directive represent a range of the VueJS Spreadsheet.
 * It must be contained in a `e-sheet` directive.
 * ```vue
 * <ejs-spreadsheet>
 *   <e-sheets>
 *    <e-sheet>
 *    <e-ranges>
 *    <e-range :dataSource='data'></e-range>
 *    </e-ranges>
 *    </e-sheet>
 *   </e-sheets>
 * </ejs-spreadsheet>
 * ```
 */
let RangeDirective = class RangeDirective extends vueImport$5 {
    render() {
        return;
    }
    getTag() {
        return 'e-range';
    }
};
RangeDirective = __decorate$5([
    EJComponentDecorator({}, isExecute$5)
], RangeDirective);
const RangePlugin = {
    name: 'e-range',
    install(Vue$$1) {
        Vue$$1.component(RangePlugin.name, RangeDirective);
    }
};

var __decorate$6 = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
const isExecute$6 = gh ? false : true;
let vueImport$6;
if (!isExecute$6 || parseInt(allVue.version) < 3) {
    vueImport$6 = Vue;
}
else {
    vueImport$6 = Vue$1;
}
let ConditionalFormatsDirective = class ConditionalFormatsDirective extends vueImport$6 {
    render() {
        return;
    }
    getTag() {
        return 'e-conditionalformats';
    }
};
ConditionalFormatsDirective = __decorate$6([
    EJComponentDecorator({}, isExecute$6)
], ConditionalFormatsDirective);
const ConditionalFormatsPlugin = {
    name: 'e-conditionalformats',
    install(Vue$$1) {
        Vue$$1.component(ConditionalFormatsPlugin.name, ConditionalFormatsDirective);
    }
};
/**
 * `e-conditionalformat` directive represent a conditionalformat of the VueJS Spreadsheet.
 * It must be contained in a `e-sheet` directive.
 * ```vue
 * <ejs-spreadsheet>
 *   <e-sheets>
 *    <e-sheet>
 *    <e-conditionalformats>
 *    <e-conditionalformat></e-conditionalformat>
 *    </e-conditionalformats>
 *    </e-sheet>
 *   </e-sheets>
 * </ejs-spreadsheet>
 * ```
 */
let ConditionalFormatDirective = class ConditionalFormatDirective extends vueImport$6 {
    render() {
        return;
    }
    getTag() {
        return 'e-conditionalformat';
    }
};
ConditionalFormatDirective = __decorate$6([
    EJComponentDecorator({}, isExecute$6)
], ConditionalFormatDirective);
const ConditionalFormatPlugin = {
    name: 'e-conditionalformat',
    install(Vue$$1) {
        Vue$$1.component(ConditionalFormatPlugin.name, ConditionalFormatDirective);
    }
};

var __decorate$7 = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
const isExecute$7 = gh ? false : true;
let vueImport$7;
if (!isExecute$7 || parseInt(allVue.version) < 3) {
    vueImport$7 = Vue;
}
else {
    vueImport$7 = Vue$1;
}
let SheetsDirective = class SheetsDirective extends vueImport$7 {
    render() {
        return;
    }
    getTag() {
        return 'e-sheets';
    }
};
SheetsDirective = __decorate$7([
    EJComponentDecorator({}, isExecute$7)
], SheetsDirective);
const SheetsPlugin = {
    name: 'e-sheets',
    install(Vue$$1) {
        Vue$$1.component(SheetsPlugin.name, SheetsDirective);
    }
};
/**
 * `e-sheet` directive represent a sheet of the VueJS Spreadsheet.
 * It must be contained in a Spreadsheet component(`ejs-spreadsheet`).
 * ```vue
 * <ejs-spreadsheet>
 *   <e-sheets>
 *    <e-sheet></e-sheet>
 *    <e-sheet></e-sheet>
 *   </e-sheets>
 * </ejs-spreadsheet>
 * ```
 */
let SheetDirective = class SheetDirective extends vueImport$7 {
    render() {
        return;
    }
    getTag() {
        return 'e-sheet';
    }
};
SheetDirective = __decorate$7([
    EJComponentDecorator({}, isExecute$7)
], SheetDirective);
const SheetPlugin = {
    name: 'e-sheet',
    install(Vue$$1) {
        Vue$$1.component(SheetPlugin.name, SheetDirective);
    }
};

var __decorate$8 = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
const isExecute$8 = gh ? false : true;
let vueImport$8;
if (!isExecute$8 || parseInt(allVue.version) < 3) {
    vueImport$8 = Vue;
}
else {
    vueImport$8 = Vue$1;
}
let DefinedNamesDirective = class DefinedNamesDirective extends vueImport$8 {
    render() {
        return;
    }
    getTag() {
        return 'e-definednames';
    }
};
DefinedNamesDirective = __decorate$8([
    EJComponentDecorator({}, isExecute$8)
], DefinedNamesDirective);
const DefinedNamesPlugin = {
    name: 'e-definednames',
    install(Vue$$1) {
        Vue$$1.component(DefinedNamesPlugin.name, DefinedNamesDirective);
    }
};
/**
 * `e-definedname` directive represent a defined name of the VueJS Spreadsheet.
 * It must be contained in a Spreadsheet component(`ejs-spreadsheet`).
 * ```vue
 * <ejs-spreadsheet>
 *   <e-definednames>
 *    <e-definedname></e-definedname>
 *    <e-definedname></e-definedname>
 *   </e-definednames>
 * </ejs-spreadsheet>
 * ```
 */
let DefinedNameDirective = class DefinedNameDirective extends vueImport$8 {
    render() {
        return;
    }
    getTag() {
        return 'e-definedname';
    }
};
DefinedNameDirective = __decorate$8([
    EJComponentDecorator({}, isExecute$8)
], DefinedNameDirective);
const DefinedNamePlugin = {
    name: 'e-definedname',
    install(Vue$$1) {
        Vue$$1.component(DefinedNamePlugin.name, DefinedNameDirective);
    }
};

var __decorate$9 = (undefined && undefined.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
// {{VueImport}}
const properties = ['activeSheetIndex', 'allowCellFormatting', 'allowChart', 'allowConditionalFormat', 'allowDataValidation', 'allowDelete', 'allowEditing', 'allowFiltering', 'allowFindAndReplace', 'allowHyperlink', 'allowImage', 'allowInsert', 'allowMerge', 'allowNumberFormatting', 'allowOpen', 'allowResizing', 'allowSave', 'allowScrolling', 'allowSorting', 'allowUndoRedo', 'allowWrap', 'cellStyle', 'cssClass', 'definedNames', 'enableClipboard', 'enableContextMenu', 'enableKeyboardNavigation', 'enableKeyboardShortcut', 'enablePersistence', 'enableRtl', 'height', 'isProtected', 'locale', 'openUrl', 'password', 'saveUrl', 'scrollSettings', 'selectionSettings', 'sheets', 'showFormulaBar', 'showRibbon', 'showSheetTabs', 'width', 'actionBegin', 'actionComplete', 'afterHyperlinkClick', 'afterHyperlinkCreate', 'beforeCellFormat', 'beforeCellRender', 'beforeCellSave', 'beforeConditionalFormat', 'beforeDataBound', 'beforeHyperlinkClick', 'beforeHyperlinkCreate', 'beforeOpen', 'beforeSave', 'beforeSelect', 'beforeSort', 'cellEdit', 'cellEditing', 'cellSave', 'contextMenuBeforeClose', 'contextMenuBeforeOpen', 'contextMenuItemSelect', 'created', 'dataBound', 'dataSourceChanged', 'dialogBeforeOpen', 'fileMenuBeforeClose', 'fileMenuBeforeOpen', 'fileMenuItemSelect', 'openComplete', 'openFailure', 'queryCellInfo', 'saveComplete', 'select', 'sortComplete'];
const modelProps = [];
const testProp = getProps({ props: properties });
const props = testProp[0];
const watch = testProp[1];
const emitProbs = Object.keys(watch);
emitProbs.push('modelchanged');
for (let props of modelProps) {
    emitProbs.push('update:' + props);
}
const isExecute$9 = gh ? false : true;
/**
 * `ejs-spreadsheet` represents the VueJS Spreadsheet Component.
 * ```vue
 * <ejs-spreadsheet></ejs-spreadsheet>
 * ```
 */
let SpreadsheetComponent = 
/* Start Options({
    props: props,
    watch: watch,
    emits: emitProbs
}) End */
class SpreadsheetComponent extends ComponentBase {
    constructor() {
        super(arguments);
        this.propKeys = properties;
        this.models = modelProps;
        this.hasChildDirective = true;
        this.hasInjectedModules = true;
        this.tagMapper = { "e-sheets": { "e-sheet": { "e-rows": { "e-row": { "e-cells": { "e-cell": { "e-images": "e-image", "e-charts": "e-chart" } } } }, "e-columns": "e-column", "e-ranges": "e-range", "e-conditionalformats": "e-conditionalformat" } }, "e-definednames": "e-definedname" };
        this.tagNameMapper = { "e-images": "e-image", "e-charts": "e-chart", "e-conditionalformats": "e-conditionalFormats", "e-definednames": "e-definedNames" };
        this.isVue3 = !isExecute$9;
        this.ej2Instances = new Spreadsheet({});
        this.bindProperties();
        this.ej2Instances._setProperties = this.ej2Instances.setProperties;
        this.ej2Instances.setProperties = this.setProperties;
    }
    setProperties(prop, muteOnChange) {
        if (this.isVue3) {
            this.models = !this.models ? this.ej2Instances.referModels : this.models;
        }
        if (this.ej2Instances && this.ej2Instances._setProperties) {
            this.ej2Instances._setProperties(prop, muteOnChange);
        }
        if (prop && this.models && this.models.length) {
            Object.keys(prop).map((key) => {
                this.models.map((model) => {
                    if ((key === model) && !(/datasource/i.test(key))) {
                        if (this.isVue3) {
                            this.ej2Instances.vueInstance.$emit('update:' + key, prop[key]);
                        }
                        else {
                            this.$emit('update:' + key, prop[key]);
                        }
                    }
                });
            });
        }
    }
    render(createElement) {
        let h = gh || createElement;
        let slots = null;
        if (!isNullOrUndefined(this.$slots.default)) {
            slots = gh ? this.$slots.default() : this.$slots.default;
        }
        return h('div', slots);
    }
    Unfreeze(sheet) {
        return this.ej2Instances.Unfreeze(sheet);
    }
    addContextMenuItems(items, text, insertAfter, isUniqueId) {
        return this.ej2Instances.addContextMenuItems(items, text, insertAfter, isUniqueId);
    }
    addCustomFunction(functionHandler, functionName) {
        return this.ej2Instances.addCustomFunction(functionHandler, functionName);
    }
    addDataValidation(rules, range) {
        return this.ej2Instances.addDataValidation(rules, range);
    }
    addDefinedName(definedName) {
        return this.ej2Instances.addDefinedName(definedName);
    }
    addFileMenuItems(items, text, insertAfter, isUniqueId) {
        return this.ej2Instances.addFileMenuItems(items, text, insertAfter, isUniqueId);
    }
    addHyperlink(hyperlink, address) {
        return this.ej2Instances.addHyperlink(hyperlink, address);
    }
    addInvalidHighlight(range) {
        return this.ej2Instances.addInvalidHighlight(range);
    }
    addRibbonTabs(items, insertBefore) {
        return this.ej2Instances.addRibbonTabs(items, insertBefore);
    }
    addToolbarItems(tab, items, index) {
        return this.ej2Instances.addToolbarItems(tab, items, index);
    }
    applyFilter(predicates, range) {
        return this.ej2Instances.applyFilter(predicates, range);
    }
    autoFit(range) {
        return this.ej2Instances.autoFit(range);
    }
    cellFormat(style, range) {
        return this.ej2Instances.cellFormat(style, range);
    }
    clear(options) {
        return this.ej2Instances.clear(options);
    }
    clearConditionalFormat(range) {
        return this.ej2Instances.clearConditionalFormat(range);
    }
    clearFilter(field) {
        return this.ej2Instances.clearFilter(field);
    }
    closeEdit() {
        return this.ej2Instances.closeEdit();
    }
    computeExpression(formula) {
        return this.ej2Instances.computeExpression(formula);
    }
    conditionalFormat(conditionalFormat) {
        return this.ej2Instances.conditionalFormat(conditionalFormat);
    }
    copy(address) {
        return this.ej2Instances.copy(address);
    }
    cut(address) {
        return this.ej2Instances.cut(address);
    }
    delete(startIndex, endIndex, model, sheet) {
        return this.ej2Instances.delete(startIndex, endIndex, model, sheet);
    }
    deleteChart(id) {
        return this.ej2Instances.deleteChart(id);
    }
    deleteImage(id, range) {
        return this.ej2Instances.deleteImage(id, range);
    }
    duplicateSheet(sheetIndex) {
        return this.ej2Instances.duplicateSheet(sheetIndex);
    }
    enableContextMenuItems(items, enable, isUniqueId) {
        return this.ej2Instances.enableContextMenuItems(items, enable, isUniqueId);
    }
    enableFileMenuItems(items, enable, isUniqueId) {
        return this.ej2Instances.enableFileMenuItems(items, enable, isUniqueId);
    }
    enableRibbonTabs(tabs, enable) {
        return this.ej2Instances.enableRibbonTabs(tabs, enable);
    }
    enableToolbarItems(tab, items, enable) {
        return this.ej2Instances.enableToolbarItems(tab, items, enable);
    }
    endEdit() {
        return this.ej2Instances.endEdit();
    }
    find(args) {
        return this.ej2Instances.find(args);
    }
    findAll(value, mode, isCSen, isEMatch, sheetIndex) {
        return this.ej2Instances.findAll(value, mode, isCSen, isEMatch, sheetIndex);
    }
    freezePanes(row, column, sheet) {
        return this.ej2Instances.freezePanes(row, column, sheet);
    }
    getData(address) {
        return this.ej2Instances.getData(address);
    }
    getDisplayText(cell) {
        return this.ej2Instances.getDisplayText(cell);
    }
    getRowData(index, sheetIndex) {
        return this.ej2Instances.getRowData(index, sheetIndex);
    }
    getSelectAllContent() {
        return this.ej2Instances.getSelectAllContent();
    }
    goTo(address) {
        return this.ej2Instances.goTo(address);
    }
    hideFileMenuItems(items, hide, isUniqueId) {
        return this.ej2Instances.hideFileMenuItems(items, hide, isUniqueId);
    }
    hideRibbonTabs(tabs, hide) {
        return this.ej2Instances.hideRibbonTabs(tabs, hide);
    }
    hideSpinner() {
        return this.ej2Instances.hideSpinner();
    }
    hideToolbarItems(tab, indexes, hide) {
        return this.ej2Instances.hideToolbarItems(tab, indexes, hide);
    }
    insertChart(chart) {
        return this.ej2Instances.insertChart(chart);
    }
    insertColumn(startColumn, endColumn, sheet) {
        return this.ej2Instances.insertColumn(startColumn, endColumn, sheet);
    }
    insertImage(images, range) {
        return this.ej2Instances.insertImage(images, range);
    }
    insertRow(startRow, endRow, sheet) {
        return this.ej2Instances.insertRow(startRow, endRow, sheet);
    }
    insertSheet(startSheet, endSheet) {
        return this.ej2Instances.insertSheet(startSheet, endSheet);
    }
    lockCells(range, isLocked) {
        return this.ej2Instances.lockCells(range, isLocked);
    }
    merge(range, type) {
        return this.ej2Instances.merge(range, type);
    }
    moveSheet(position, sheetIndexes) {
        return this.ej2Instances.moveSheet(position, sheetIndexes);
    }
    numberFormat(format, range) {
        return this.ej2Instances.numberFormat(format, range);
    }
    open(options) {
        return this.ej2Instances.open(options);
    }
    openFromJson(options) {
        return this.ej2Instances.openFromJson(options);
    }
    paste(address, type) {
        return this.ej2Instances.paste(address, type);
    }
    protectSheet(sheet, protectSettings) {
        return this.ej2Instances.protectSheet(sheet, protectSettings);
    }
    redo() {
        return this.ej2Instances.redo();
    }
    refresh(isNew) {
        return this.ej2Instances.refresh(isNew);
    }
    removeContextMenuItems(items, isUniqueId) {
        return this.ej2Instances.removeContextMenuItems(items, isUniqueId);
    }
    removeDataValidation(range) {
        return this.ej2Instances.removeDataValidation(range);
    }
    removeDefinedName(definedName, scope) {
        return this.ej2Instances.removeDefinedName(definedName, scope);
    }
    removeHyperlink(range) {
        return this.ej2Instances.removeHyperlink(range);
    }
    removeInvalidHighlight(range) {
        return this.ej2Instances.removeInvalidHighlight(range);
    }
    replace(args) {
        return this.ej2Instances.replace(args);
    }
    resize() {
        return this.ej2Instances.resize();
    }
    save(saveOptions) {
        return this.ej2Instances.save(saveOptions);
    }
    saveAsJson() {
        return this.ej2Instances.saveAsJson();
    }
    selectRange(address) {
        return this.ej2Instances.selectRange(address);
    }
    setBorder(style, range, type) {
        return this.ej2Instances.setBorder(style, range, type);
    }
    setColWidth(width, colIndex, sheetIndex) {
        return this.ej2Instances.setColWidth(width, colIndex, sheetIndex);
    }
    setRowHeight(height, rowIndex, sheetIndex, edited) {
        return this.ej2Instances.setRowHeight(height, rowIndex, sheetIndex, edited);
    }
    showSpinner() {
        return this.ej2Instances.showSpinner();
    }
    sort(sortOptions, range) {
        return this.ej2Instances.sort(sortOptions, range);
    }
    startEdit() {
        return this.ej2Instances.startEdit();
    }
    undo() {
        return this.ej2Instances.undo();
    }
    unprotectSheet(sheet) {
        return this.ej2Instances.unprotectSheet(sheet);
    }
    updateAction(options) {
        return this.ej2Instances.updateAction(options);
    }
    updateCell(cell, address) {
        return this.ej2Instances.updateCell(cell, address);
    }
    updateUndoRedoCollection(args) {
        return this.ej2Instances.updateUndoRedoCollection(args);
    }
    wrap(address, wrap) {
        return this.ej2Instances.wrap(address, wrap);
    }
};
SpreadsheetComponent = __decorate$9([
    EJComponentDecorator({
        props: properties
    }, isExecute$9)
    /* Start Options({
        props: props,
        watch: watch,
        emits: emitProbs
    }) End */
], SpreadsheetComponent);
const SpreadsheetPlugin = {
    name: 'ejs-spreadsheet',
    install(Vue$$1) {
        Vue$$1.component(SpreadsheetPlugin.name, SpreadsheetComponent);
        Vue$$1.component(SheetPlugin.name, SheetDirective);
        Vue$$1.component(SheetsPlugin.name, SheetsDirective);
        Vue$$1.component(RowPlugin.name, RowDirective);
        Vue$$1.component(RowsPlugin.name, RowsDirective);
        Vue$$1.component(CellPlugin.name, CellDirective);
        Vue$$1.component(CellsPlugin.name, CellsDirective);
        Vue$$1.component(ImagePlugin.name, ImageDirective);
        Vue$$1.component(ImagesPlugin.name, ImagesDirective);
        Vue$$1.component(ChartPlugin.name, ChartDirective);
        Vue$$1.component(ChartsPlugin.name, ChartsDirective);
        Vue$$1.component(ColumnPlugin.name, ColumnDirective);
        Vue$$1.component(ColumnsPlugin.name, ColumnsDirective);
        Vue$$1.component(RangePlugin.name, RangeDirective);
        Vue$$1.component(RangesPlugin.name, RangesDirective);
        Vue$$1.component(ConditionalFormatPlugin.name, ConditionalFormatDirective);
        Vue$$1.component(ConditionalFormatsPlugin.name, ConditionalFormatsDirective);
        Vue$$1.component(DefinedNamePlugin.name, DefinedNameDirective);
        Vue$$1.component(DefinedNamesPlugin.name, DefinedNamesDirective);
    }
};

export { ImagesDirective, ImageDirective, ImagesPlugin, ImagePlugin, ChartsDirective, ChartDirective, ChartsPlugin, ChartPlugin, CellsDirective, CellDirective, CellsPlugin, CellPlugin, RowsDirective, RowDirective, RowsPlugin, RowPlugin, ColumnsDirective, ColumnDirective, ColumnsPlugin, ColumnPlugin, RangesDirective, RangeDirective, RangesPlugin, RangePlugin, ConditionalFormatsDirective, ConditionalFormatDirective, ConditionalFormatsPlugin, ConditionalFormatPlugin, SheetsDirective, SheetDirective, SheetsPlugin, SheetPlugin, DefinedNamesDirective, DefinedNameDirective, DefinedNamesPlugin, DefinedNamePlugin, SpreadsheetComponent, SpreadsheetPlugin };
export * from '@syncfusion/ej2-spreadsheet';
